### PR TITLE
Return 422 status when login fails + treat `turbo_stream` as a navigational format

### PR DESCRIPTION
**Pull request by ghiculescu**
Devise `FailureApp` now returns a `422` status when running the `recall` flow. This, combined with https://github.com/heartcombo/responders/pull/223, enables Devise to work correctly with the [new Rails defaults](https://github.com/rails/rails/pull/41026) for form errors, as well as with new libraries like Turbo.

By default, the `recall` flow only runs on the [`SessionsController`](https://github.com/heartcombo/devise/blob/v4.7.3/app/controllers/devise/sessions_controller.rb#L48).

`:turbo_stream` is also treated as a navigational format which means Devise will return redirects where appropriate rather than `401`s for Turbo requests.

Fixes https://github.com/heartcombo/devise/pull/5325

### CODE REVIEW

**Review by santib — APPROVED** [review]
Looks good to me. Should this PR also upgrade the responders gem to the version that responds 422 on error?

**Comment by ghiculescu** [comment]
Yes, it should. Once there's a new release of `responders` I can do that. The last release was a few weeks ago https://github.com/heartcombo/responders/releases

**Comment by carlosantoniodasilva** [comment]
Thanks @ghiculescu, I'll take a closer look in the next few days. And I have yet to release a new responders version with that change, hopefully coming soon to a rubygems near you. :)

**Comment by ghiculescu** [comment]
All those test fails are Docker errors 😿

**Review by marivaldo — APPROVED** [review]

**Comment by ghiculescu** [comment]
@carlosantoniodasilva friendly bump here, anything I can do to help get new versions of Devise + Responders with these improvements shipped?

**Comment by Petercopter** [comment]
I've tested these changes locally with `hotwire-rails`.
* I'm getting the 422 when session creation fails
* Seeing the flash message
* Logs are reporting `Processing by SessionsController#new as TURBO_STREAM`

Looks good to me!
 
Thanks to everyone I've been following along with while we get Hotwire going! 👍  🎉

**Comment by carlosantoniodasilva** [comment]
@ghiculescu thanks. Aside from testing this myself, I need to think some more about the breaking change and how this can affect other apps, and a game plan for releasing it... I'm not sure Devise is ready for a major bump, without other things being considered, and I'm not sure we actually need one yet. Anyway, this is all on me now, and I'm working through it.

@Petercopter thanks for testing & confirming this works for you as well.

**Comment by jasonfb** [comment]
this isn't released yet is it?

I just found this (through debugging) and filed this on Stack Overflow (mostly for the next person)

https://stackoverflow.com/questions/66615478/turbo-rails-with-devise-does-not-redirect-consistently-rails-6-1-3-devise-4-7-3

My reproduction app against Devise can be found here:
https://github.com/jasonfb/TR001

I'd love to be able to upgrade my Rails 6.1 app for Turbo-Rails today, but my upgrade is riddled with these Devise bugs-- mostly redirects that don't redirect in the browser but the action has happened on the backend.
 
I see this pull appears to fix everything (🎉 !), true? false? truthy? falsy? 

in master soon? Should I wait a bit? I guess if it were in master I could point my gem to the master branch of devise? Any tips appreciated, even just a monkey patch to get me going until the fix is released.

**Comment by carlosantoniodasilva** [comment]
@jasonfb Devise stable (4.7.3) or master do not support Turbo fully, this PR is aimed to do that, you can try pointing to the PR's branch and test it out yourself to see how that works. I cannot guarantee or give an ETA about a Devise stable release that supports Turbo yet.

And FWIW, I don't consider this a bug per-se, but a change driven by a new library out there that Devise can accommodate to work better out of the box.

**Comment by tmaier** [comment]
@jasonfb maybe this screencast helps you: https://gorails.com/episodes/devise-hotwire-turbo

**Comment by ghiculescu** [comment]
FYI: if you want to use this for now, you can put this in your Gemfile:

```ruby
gem "devise", github: "ghiculescu/devise", branch: "error-code-422" # https://github.com/heartcombo/devise/pull/5340 not yet merged
gem "responders", github: "heartcombo/responders" # https://github.com/heartcombo/responders/pull/223 not yet released
```

**Comment by adrianvalenz** [comment]
> FYI: if you want to use this for now, you can put this in your Gemfile:
> 
> ```ruby
> gem "devise", github: "ghiculescu/devise", branch: "error-code-422" # https://github.com/heartcombo/devise/pull/5340 not yet merged
> gem "responders", github: "heartcombo/responders" # https://github.com/heartcombo/responders/pull/223 not yet released
> ```

Thank you for this. I just added these two gems and devise/hotwire working seamlessly together. There was no need to edit any file.

**Comment by JoeWoodward** [comment]
> @ghiculescu thanks. Aside from testing this myself, I need to think some more about the breaking change and how this can affect other apps, and a game plan for releasing it... I'm not sure Devise is ready for a major bump, without other things being considered, and I'm not sure we actually need one yet. Anyway, this is all on me now, and I'm working through it.
> 
> @Petercopter thanks for testing & confirming this works for you as well.

@carlosantoniodasilva are there any plans for getting this PR merged at some point? I'm currently disabling turbo on all my devise forms but would really love to let turbo work its magic. If there's a plan/timeline on when this PR will be merged I'll happily wait, otherwise I'll go ahead and use the PR branch, however that is out of date now so I'll have to fork and rebase so would rather just use devise if a release is not too far off.  
We're eager to use Turbo and with Devise being a staple in the Rails ecosphere it is putting us (my company) off upgrading all of our Rails apps to Turbo and I'm sure others too.

Thank you for your work. Not trying to pressure you, just looking for some insight so I can take appropriate actions.

**Comment by silva96** [comment]
Hi, when is this going to be released?

**Comment by archonic** [comment]
I've also tested this (via [@ghiculescu's comment](https://github.com/heartcombo/devise/pull/5340#issuecomment-833840004)) and it fixed the 422 rendering issue when submitting an erroneous form.

It doesn't seem to solve all turbo compatibility however. When submitting a valid form, the line `respond_with resource, location: after_inactive_sign_up_path_for(resource)` will render a 200. In that case the client is just left looking at the submitted form.

**Comment by archonic** [comment]
I'm still working out what the issue is to be honest. A redirect also doesn't render on the client even though the server is rendering the redirect and the page that is the redirect location.

**Comment by yrashk** [comment]
@ghiculescu @archonic I've used @ghiculescu's branch and it seems to have worked fine for me. I do use registrations and I run system tests on this and I tried it manually as well.

I use a copy of @ghiculescu's branch, rebased on top of 4.8.0: https://github.com/hackerintro/devise/tree/error-code-422

It's not in production yet, but if there's indeed a problem, it'd be great to figure out what it is before it'll be deployed. I am happy to help working this out, @ghiculescu @archonic

**Comment by archonic** [comment]
The issue I was running into was something else. I was trying to upgrade from Turbolinks to Turbo in my main app. I still don't know what the issue was but it's explained [here](https://stackoverflow.com/questions/68386745/turbolinks-to-turbo-upgrade-has-broken-form-redirection). I've since given up on trying to implement Turbo beta in my main app. I'll have to come up with a smaller reproduction at some point.

**Comment by yrashk** [comment]
I also had some issues upgrading to Turbo, but I wasn't trying to upgrade Devise to Turbo at that point. Once I've done that, I used the patch from that branch and it just worked. Happy to help figuring your issue out so we can perhaps help Devise get up-to-date with Turbo at last.

**Comment by miharekar** [comment]
So…what's happening with this? 😅

**Comment by brunoprietog** [comment]
Any news on this?

cc @carlosantoniodasilva

Thanks!

**Comment by Matt-Yorkley** [comment]
Hey @carlosantoniodasilva is there anything that can be done to help move this along? Any Rails apps using Turbo Streams (or [MRUJS](https://mrujs.com) for that matter) for handling form submissions are basically broken when it comes to Devise.

Is the fear of potential disruption for non-Turbo users the main thing holding this back? Would something like an opt-in feature toggle for the new behavior be a more acceptable path forward WRT the change in status code in the responses?

Basically I'd be over the moon to see a Devise release with this PR in it...

:pray:

**Comment by carlosantoniodasilva** [comment]
Appreciate the ping. I'm well aware of the Rails 7 release timeline (final is still weeks/months away) and that Devise is currently not working with it at all, this being part of the necessary changes. I was waiting for things to play out further before making any changes to Responders or Devise to accommodate Rails master, plus I haven't had much time and energy to be honest, and my first attempts to test Responders alone weren't fully working with just the changes here (which might just mean I wasn't testing things right). In any case, now that the alpha was released it does make it a bit more pressing, I'll find some time to get things lined up on this side soon. Thanks.

**Comment by saeedjabbar** [comment]
Why not merge the fix as it will be a while before many upgrade to Rails 7?
Let us know if there's anything the community can help with.

**Comment by giorgenes** [comment]
I've just installed rails 7 + devise and ran into this issue as well.

**Comment by manuelreis** [comment]
Hi, are there plans to merge this PR or an alternative approach is going to be followed?

**Comment by ghiculescu** [comment]
anyone know what's up with those tests?

**Comment by baarkerlounger** [comment]
We've been testing this branch successfully so far, just two small additional fixes required that we've come across so far, in case they're generally helpful:

https://github.com/baarkerlounger/devise/commit/223969286e51b02e78309b7dad908d223d6ba7f5
https://github.com/baarkerlounger/devise/commit/ac956cc2a58daa57e071202967a212a1dac08053

**Comment by ghiculescu** [comment]
Thanks @baarkerlounger ! I've pushed those commits to this branch, with you as the co-author. Please let me know if there's any extra tests / fixes worth adding here - if you push them to your branch I can cherry-pick them here.

@jasonfb maybe? I'd be (pleasantly!) surprised if this PR is no longer needed. It'd be helpful to know what Rails / Devise / Turbo versions you are using.

**Comment by baarkerlounger** [comment]
@ghiculescu https://github.com/baarkerlounger/devise/commit/6d0b6b52a9d0e87ae6d9f9acb562169751623078

**Comment by ghiculescu** [comment]
I realise this might be a hard ask, but is it possible to share a repo that I can reproduce the issue with? It's hard to guess what's going on just from the gifs/logs.

**Comment by scarroll32** [comment]
Any update on this?  Is it still needed or not?

**Comment by jasonfb** [comment]
https://github.com/heartcombo/devise/issues/5478

**Comment by dnalbach** [comment]
@carlosantoniodasilva - thank you for all your work maintaining this project - you mentioned the following six months ago:

> That said, I definitely have plans to work on making it fully Turbo-aware/compatible before 7 final is out. Thanks.

Rails 7 is out now and it looks like the Turbo support hasn't fully arrived yet. Life happens, and I've got a family, so I understand.

Is there anything that you need to help this along? Is it a money issue, a time issue, a design issue, or ?

I may be able to help with any or all of those including a bounty for Turbo support completion. 

Is there somewhere that a coordinated group effort is taking place for Devise Turbo support that I could join and help? Please reach out to me if you are able to at:

daniel@fullstackerconsulting.com

**Comment by joemasilotti** [comment]
I'm also happy to help contribute if that's a blocker. I didn't see a Sponsor button anywhere on the repo, though.

**Comment by maxwell** [comment]
Wanted to flag this PR in `turbo-rails` https://github.com/hotwired/turbo-rails/pull/367

If we had this merged in, I think it would be a useful tool in Devise fully supporting turbo, specifically making a bit nicer when linking to a resource wrapped in an auth check, and maintaining devise's default behavior of redirecting to sign in and restoring the location afterwords.

**Comment by pepper-roni** [comment]
Hey! I just ran into this issue! This seems to solve it, but its been sitting here. It sounds like the block is breaking support for non-turbo rails? Could we at least add `data: {turbo: false}` in another pr? Or even some info in the readme on state of devise and turbo and workarounds. Happy to help test or write things up to get it moving. For now ill run on this branch... thanks @ghiculescu  !

**Comment by 1klap** [comment]
i'm running this branch on production for the moment and have no issues with it. i'd love ofc to go back to a devise release as soon as this is shipped. can i help with anything? 

i understand that is can be considered a breaking change, could it be extracted into an opt-in config?

File: app/controllers/devise/passwords_controller.rb [hunk 1]
```diff
@@ -47,7 +47,7 @@ def update
       respond_with resource, location: after_resetting_password_path_for(resource)
     else
       set_minimum_password_length
-      respond_with resource
+      respond_with resource, status: :unprocessable_entity
```
Comment (carlosantoniodasilva) [inline app/controllers/devise/passwords_controller.rb:50]: Were you having any issue with responders alone here, to have to enforce a status?

File: app/controllers/devise/sessions_controller.rb [hunk 2]
```diff
@@ -77,7 +77,7 @@ def respond_to_on_destroy
     # support returning empty response on GET request
     respond_to do |format|
       format.all { head :no_content }
-      format.any(*navigational_formats) { redirect_to after_sign_out_path_for(resource_name) }
+      format.any(*navigational_formats) { redirect_to after_sign_out_path_for(resource_name), status: :see_other }
```
Comment (carlosantoniodasilva) [inline app/controllers/devise/sessions_controller.rb:80]: Note: responders now has a configuration `redirect_status` that allows us to override this across the board.

This status is also going to be necessary on the registrations controller destroy action:
https://github.com/heartcombo/devise/blob/3632ddf67473880ed9583201ce13f108c620514f/app/controllers/devise/registrations_controller.rb#L70

File: lib/devise/failure_app.rb [hunk 3]
```diff
@@ -71,7 +71,9 @@ def recall
       end
 
       flash.now[:alert] = i18n_message(:invalid) if is_flashing_format?
-      self.response = recall_app(warden_options[:recall]).call(request.env)
+      response_from_app = recall_app(warden_options[:recall]).call(request.env)
+      response_from_app[0] = recall_response_code(response_from_app[0])
+      self.response = response_from_app
```
Comment (carlosantoniodasilva) [inline lib/devise/failure_app.rb:76]: :+1: this is kind of what I had in mind too, change the status code here for failed login attempts.

I thought we could use responders though since we now have this configured there, so I went with that. This way, it can be overridden across the board and not only for this particular response.

**Review by carlosantoniodasilva — COMMENTED** [review]
Hey @ghiculescu, appreciate our work here, and sorry again for such a long time replying back.

I left a few thoughts below as I looked through the changes, but no need to change anything yourself here. I've been working on related changes https://github.com/heartcombo/devise/pull/5548 on top of what you did here and a few other PRs/issues, to try to accommodate for turbo without making it a full breaking change, and integrating with responders. Your updates here were extremely helpful to get a better understanding of what was needed, and get that up and running, so thank you.

Please anyone feel free to give that one a shot, and leave any feedback or issue you may encounter there. Let's also keep this open for now, it can be closed if / when that one gets merged. I still have some bits to work out, and some tests to add/update, but the bulk of the code should be there. Thanks again.

File: app/controllers/devise/passwords_controller.rb (hunk 1 above)
Comment (ghiculescu) [reply app/controllers/devise/passwords_controller.rb:50]: @baarkerlounger made that change in https://github.com/ghiculescu/devise/commit/eef20a326075b3f36df7f8b2a35c6f5ebe4b4f6b

The discussion is here https://github.com/heartcombo/devise/pull/5340#issuecomment-1029004041, I don't really remember any context beyond that.

Comment (carlosantoniodasilva) [reply app/controllers/devise/passwords_controller.rb:50]: Gotcha, thanks. I think those should be handled by default with responders now :+1:

**Comment by carlosantoniodasilva** [comment]
The main branch should contain all that's necessary for fully working with Turbo now, based on the changes here and a few others. A new version will be released soon, but feel free to test it out from the main branch in the meantime, and report back on any issues. Thanks.
